KAFKA-5967 Ineffective check of negative value in CompositeReadOnlyKeyValueStore#approximateNumEntries()

shivsantham opened the pull request:
package name: org.apache.kafka.streams.state.internals
Minor change to approximateNumEntries() method in CompositeReadOnlyKeyValueStore class.

long total = 0;
   for (ReadOnlyKeyValueStore<K, V> store : stores) {
          total += store.approximateNumEntries();
   }

return total < 0 ? Long.MAX_VALUE : total;

The check for negative value seems to account for wrapping. However, wrapping can happen within the for loop. So the check should be performed inside the loop.

ssanthalingam committed:
KAFKA-5967 Ineffective check of negative value in CompositeReadOnlyKeyValueStore#approximateNumEntries()

long total = 0;
   for (ReadOnlyKeyValueStore<K, V> store : stores) {
          total += store.approximateNumEntries();
   }

return total < 0 ? Long.MAX_VALUE : total;

The check for negative value seems to account for wrapping. However, wrapping can happen within the for loop. So the check should be performed inside the loop.

tedyu commented:
lgtm

mjsax reviewed (commented):
Change LGTM. Can you please add a test? Thx.

ssanthalingam committed:
KAFKA-5967 Ineffective check of negative value in CompositeReadOnlyKeyValueStore

Adding a test for #KAFKA-5967

shivsantham commented:
@mjsax  Added a test. Please let me know if that is fine.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyKeyValueStoreTest.java:
```diff
@@ -257,6 +257,19 @@ public long approximateNumEntries() {
+            }
+        });
+
+        stubOneUnderlying.put("underflow", "me");
```
I this really needed?

shivsantham replied:
Got rid of this.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/CompositeReadOnlyKeyValueStoreTest.java:
```diff
@@ -257,6 +257,19 @@ public long approximateNumEntries() {
+        stubProviderTwo.addStore(storeName, new NoOpReadOnlyStore<Object, Object>() {
+            @Override
+            public long approximateNumEntries() {
+                return Long.MIN_VALUE;
```
I would return `MAX_VALUE` here -- we should also add multiple stores that do return `MAX_VALUE` to test the inner part of the loop.

shivsantham replied:
Thanks will take care!

mjsax reviewed (commented):
Thanks for adding the test. Have some suggestions.

ssanthalingam committed:
KAFKA-5967 Ineffective check of negative value in CompositeReadOnlyKeyValueStore

-Fixing test

shivsantham commented:
@mjsax  Thank you for your suggestions. I have made the changes. Kindly review!

mjsax reviewed (approved):
LGTM. Call for final review and merging @guozhangwang @dguy

mjsax commented:
Please cherry pick to `0.11.0`, too.

ssanthalingam committed:
Flatten SMT does not work with null values

A bug in Flatten SMT while doing tests with different SMTs that are provided out-of-box. Flatten SMT does not work as expected with schemaless JSON that has properties with null values.

Example json:
  {A={D=dValue, B=null, C=cValue}}
The issue is in if statement that checks for null value.

CURRENT VERSION:
  for (Map.Entry<String, Object> entry : originalRecord.entrySet()) {
            final String fieldName = fieldName(fieldNamePrefix, entry.getKey());
            Object value = entry.getValue();
            if (value == null) {
                newRecord.put(fieldName(fieldNamePrefix, entry.getKey()), null);
                return;
            }

PROPOSED VERSION:
  for (Map.Entry<String, Object> entry : originalRecord.entrySet()) {
            final String fieldName = fieldName(fieldNamePrefix, entry.getKey());
            Object value = entry.getValue();
            if (value == null) {
                newRecord.put(fieldName(fieldNamePrefix, entry.getKey()), null);
                continue;
            }

mjsax commented:
@shivsantham  The last commit does not seem to belong to this PR...

ssanthalingam committed:
Revert "Flatten SMT does not work with null values"

This reverts commit 48e50992139c03099b3249195efc316a84d6bba1.

shivsantham commented:
Sorry i reverted the last commit that does not related to the current KAFKA bug. i will send out a seperate PR for the other one.

dguy reviewed (approved):
Thanks @shivsantham, LGTM

dguy commented:
Merged to trunk and cherry-picked to 0.11.0
